packages/x-data-grid-pro/src/hooks/features/serverSideLazyLoader/useGridDataSourceLazyLoader.ts

```diff
@@ -0,0 +1,204 @@
+  const filterModel = useGridSelector(privateApiRef, gridFilterModelSelector);
+  const paginationModel = useGridSelector(privateApiRef, gridPaginationModelSelector);
+  const renderedRowsIntervalCache = React.useRef(INTERVAL_CACHE_INITIAL_STATE);
+  const isDisabled = !props.unstable_dataSource || props.lazyLoading !== true;
+
+  const addSkeletonRows = React.useCallback(() => {
+    if (isDisabled) {
+      return;
+    }
+
+    const tree = privateApiRef.current.state.rows.tree;
+    const rootGroup = tree[GRID_ROOT_GROUP_ID] as GridGroupNode;
+    const rootGroupChildren = [...rootGroup.children];
+
+    const pageRowCount = privateApiRef.current.state.pagination.rowCount;
+    const rowCount = privateApiRef.current.getRowsCount();
+
+    for (let i = 0; i < pageRowCount - rowCount; i += 1) {
+      const skeletonId = getSkeletonRowId(i);
+
+      rootGroupChildren.push(skeletonId);
+
+      const skeletonRowNode: GridSkeletonRowNode = {
+        type: 'skeletonRow',
+        id: skeletonId,
+        parent: GRID_ROOT_GROUP_ID,
+        depth: 0,
+      };
+
+      tree[skeletonId] = skeletonRowNode;
+    }
+
+    tree[GRID_ROOT_GROUP_ID] = { ...rootGroup, children: rootGroupChildren };
+
+    privateApiRef.current.setState(
+      (state) => ({
+        ...state,
+        rows: {
+          ...state.rows,
+          tree,
+        },
+      }),
+      'addSkeletonRows',
+    );
+    privateApiRef.current.requestPipeProcessorsApplication('hydrateRows');
+  }, [privateApiRef, isDisabled]);
+
+  const handleRenderedRowsIntervalChange = React.useCallback<
+    GridEventListener<'renderedRowsIntervalChange'>
+  >(
+    (params) => {
+      if (isDisabled) {
+        return;
+      }
+
+      const getRowsParams: GridGetRowsParams = {
+        start: params.firstRowIndex,
+        end: params.lastRowIndex,
+        sortModel,
+        filterModel,
+      };
+
+      if (
+        renderedRowsIntervalCache.current.firstRowToRender === params.firstRowIndex &&
+        renderedRowsIntervalCache.current.lastRowToRender === params.lastRowIndex
+      ) {
+        return;
+      }
+
+      renderedRowsIntervalCache.current = {
+        firstRowToRender: params.firstRowIndex,
+        lastRowToRender: params.lastRowIndex,
+      };
+
+      const currentVisibleRows = getVisibleRows(privateApiRef, {
+        pagination: props.pagination,
+        paginationMode: props.paginationMode,
+      });
+
+      const skeletonRowsSection = findSkeletonRowsSection({
+        apiRef: privateApiRef,
+        visibleRows: currentVisibleRows.rows,
+        range: {
+          firstRowIndex: params.firstRowIndex,
+          lastRowIndex: params.lastRowIndex,
+        },
+      });
+
+      if (!skeletonRowsSection) {
+        return;
+      }
+
+      getRowsParams.start = skeletonRowsSection.firstRowIndex;
+      getRowsParams.end = skeletonRowsSection.lastRowIndex;
+
+      privateApiRef.current.publishEvent('getRows', getRowsParams);
```

Comment:
![image](https://github.com/user-attachments/assets/fcf8cc5b-fcba-45bd-a379-acf438f933e1)

Here are the requested values for `firstRowToRender` and `lastRowToRender`. I noticed the requests being made on irregular batch sizes.

Should we honor the `paginationModel.pageSize` here (or perhaps another argument of the data source interface like `batchSize`)? I'm thinking in terms of:

### Batch size

Some users may want to use fixed batch sizes. It could be important for some implementations where the users want to bring in the data in a defined batch size.

### Network burden

Too frequent API calls could add more burden to the backend server. Some users may want to optimize the network bandwidth by reducing the API calls sent to the server. Especially, if the backend call is expensive.

The users may choose to fetch data in batches of 100 rows each, to smooth the actual rendering of data, which would then work just like client-side data for the data already fetched.

### Caching

Since, the values of `firstRowToRender` and `lastRowToRender` somehow relate to the current viewport, which could be anything at a given point. The caching might not perform as expected here.

Imaging fetching a batch of rows of let's say these values:
```tsx
{ firstRowToRender: 10, lastRowToRender: 19 }
```

Which provides the rows on page 2 (if used with pagination with pageSize = 10).

If the rows get invalidated and the user scrolls to the same viewport again, the values to fetch could be different. Let's say:
```tsx
{ firstRowToRender: 11, lastRowToRender: 15 }
```

In this case, although we have the rows in the cache, we still can't use them due to the cache key being different. So we could have multiple copies of the same rowset in the cache, yet still not be able to use them.

### Possible solution

We could have an algorithm computing the nearest `start` and `end` values based on  `renderedRowsIntervalChange` params respecting the `paginationModel.pageSize` to determine the batch size.

For example, consider `pageSize/batchSize=10`:
1. On page load, the Grid requests data for page 1 (start: 0, end: 9)
2. When the user scrolls down, the Grid can smartly decide if the range defined in `{ firstRowIndex, lastRowIndex }` comes in the already fetched data
   - Yes: Get the specific page (or pages) of rows from the cache, extract the subset requested indexes passed by `renderedRowsIntervalChange`, and set them (let's say using the API method `replaceRows`)
   - No: Find the nearest unfetched page for the requested range, initiate a fetch, on rows fetch, do step a
3. The same would happen on `filterModel`, `sortModel` update. Based on the last rendered indexes returned by `renderedRowsIntervalChange` the grid will compute which page to fetch (and probably a few extra ones to pre-render some rows on top or bottom) and send a fetch request.

What do you think?

Reply:
I was actually thinking more or less the same
the 'improve caching' checkbox was meant to adjust start and end to have a better cache hit chance. I was thinking at first to use viewport size to determine the pagesize, but I think it is better to just rely on the pagination model. That way users can control the portion of the data that is loaded on each request.

So, fully agree with the comment above

Reply:
I realized now that, in case when you scroll to the area between two pages, two requests must be made to fill each part of the viewport

Reply:
> I realized now that, in case when you scroll to the area between two pages, two requests must be made to fill each part of the viewport.

Could we combine more requests into one in such cases? Like for example if a viewport requires rows of page 1, 2, and 3, we could set the start and end values like: `{ start: 0, end: 30}` (pageSize: 10)

Reply:
I will combine the requests, but I will keep the pages separate in the cache to have better cache hit ratio